File: leetcode/101-200/187-repeated-dna-sequences/bench/repeated_dna.py
"""Benchmark workload for LeetCode #187 — Repeated DNA Sequences (Python; scale lane)."""

import sys


def scan(bases, stamp, cnt, n, pass_id):
    mask = 1048575  # 2^20 - 1
    hash_ = 0
    dups = 0
    for i in range(n):
        hash_ = ((hash_ << 2) | bases[i]) & mask
        if i >= 9:
            key = hash_
            if stamp[key] != pass_id:
                stamp[key] = pass_id
                cnt[key] = 0
            c = cnt[key] + 1
            cnt[key] = c
            if c == 2:
                dups += 1
    return dups


def main():
    n = 1000000
    passes = 30
    tablesize = 1048576  # 2^20

    bases = [0] * n
    state = 12345
    for b in range(n):
        state = (state * 1103515245 + 12345) & 2147483647
        bases[b] = (state >> 16) % 4

    stamp = [-1] * tablesize
    cnt = [0] * tablesize

    sink = 0
    for p in range(passes):
        state = (state * 1103515245 + 12345) & 2147483647
        idx = state % n
        bases[idx] = (state >> 16) % 4
        sink += scan(bases, stamp, cnt, n, p)
    print(sink)


main()

File: leetcode/101-200/187-repeated-dna-sequences/bench/repeated_dna.rs
fn scan(bases: &[i64], stamp: &mut [i64], cnt: &mut [i64], n: i64, pass: i64) -> i64 {
    let mask: i64 = 1048575; // 2^20 - 1
    let mut hash: i64 = 0;
    let mut dups: i64 = 0;
    let mut i: i64 = 0;
    while i < n {
        hash = ((hash << 2) | bases[i as usize]) & mask;
        if i >= 9 {
            let key = hash as usize;
            if stamp[key] != pass {
                stamp[key] = pass;
                cnt[key] = 0;
            }
            let c = cnt[key] + 1;
            cnt[key] = c;
            if c == 2 {
                dups += 1;
            }
        }
        i += 1;
    }
    dups
}

fn main() {
    let n: i64 = 1000000;
    let passes: i64 = 30;
    let tablesize: i64 = 1048576; // 2^20

    let mut bases: Vec<i64> = Vec::with_capacity(n as usize);
    let mut state: i64 = 12345;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        bases.push((state >> 16) % 4);
    }

    let mut stamp: Vec<i64> = vec![-1; tablesize as usize];
    let mut cnt: Vec<i64> = vec![0; tablesize as usize];

    let mut sink: i64 = 0;
    for p in 0..passes {
        state = (state * 1103515245 + 12345) & 2147483647;
        let idx = (state % n) as usize;
        bases[idx] = (state >> 16) % 4;
        sink += scan(&bases, &mut stamp, &mut cnt, n, p);
    }
    println!("{}", sink);
}
